### PR TITLE
fix #243 ime input support

### LINKED ISSUE #243: IME Support
Hi there~

Awesome NeoVide is, though some support of non-English inputs may face some degrees of disability. Dynamic represent [here](https://twitter.com/_kjhl/status/1248993062837039105). Besides, thanks Kethku for friendly reply on my previous tweets, such a warm action.

- **Catch Tapping Events** When user is typing in non-English context, Vide catchs the letter and send it to inside vim. In normal situations with modern IDE, problem above will be (probably I guess) solved with `if_not_us_keyboard_current -> act nothing`.
- **Position of Popup Vocab List** As shown in tweets, the popup input methods was fixed at TL, something like `(0, 0)`.
- **The Convert of Chinese** Seemingly strings of pronunciation always goes under a process of converting to square Chinese chars in my terminal-based vim, while Vide does not.

I've noticed that some similar bugs were tagged but I hope this helps, at least the former two (I guess) will be a common faults in **Input Methods Situation**.

### PR DESCRIPTION
**Pull request by vzex**
fix #243

no breaking changes
no breaking build, tested on windows 10.
chineseinput

### CODE REVIEW

**Comment by vzex** [comment]
btw,this code
image
i don't konw the purpose, this will cause the editor received more string than the normal ime input, this is wrong on mac.

File: src/window/window_wrapper/keyboard_manager.rs [hunk 1]
```diff
@@ -8,14 +8,18 @@ use crate::channel_utils::LoggingTx;
 use crate::settings::SETTINGS;
 use crate::window::KeyboardSettings;
 
+enum KeyOrIme {
```
Comment (last-partizan) [inline src/window/window_wrapper/keyboard_manager.rs:11]: Maybe this should be named in some generic way, like KeyboardEvent, or InputEvent?

File: src/window/window_wrapper/keyboard_manager.rs [hunk 2]
```diff
@@ -8,14 +8,18 @@ use crate::channel_utils::LoggingTx;
...
+}
 pub struct KeyboardManager {
     command_sender: LoggingTx<UiCommand>,
     shift: bool,
     ctrl: bool,
     alt: bool,
     logo: bool,
     ignore_input_this_frame: bool,
-    queued_key_events: Vec<KeyEvent>,
+    queued_key_or_ime_events: Vec<KeyOrIme>,
```
Comment (last-partizan) [inline src/window/window_wrapper/keyboard_manager.rs:22]: And, after renaming enum, this should also be renamed to something generic.

**Review by last-partizan — APPROVED** [review]
Looks good to me.

**Comment by last-partizan** [comment]
@clason @j4qfrost @LoipesMas @smolck please tale a look, and if you don't see any problems, i'll merge this.

**Comment by last-partizan** [comment]
@vzex looks like you need to sync changes with latest main branch, to fix those failing checks.

**Comment by vzex** [comment]
> @vzex looks like you need to sync changes with latest main branch, to fix those failing checks.

done

**Review by clason — CHANGES_REQUESTED** [review]
I'm afraid this has broken dead keys (at least) on macOS: typing `´e` now inserts `éé` instead of a single `é` as on main (and similarly for other dead key combinations).

Also, typing `´´` (or other dead keys) makes dead keys "sticky", so any other dead key is inserted literally _twice_, and this continues while typing (other) dead keys.

**Comment by LoipesMas** [comment]
Looks good to me, but I didn't have any issues before. At least I don't see any regression (on Linux).

**Comment by vzex** [comment]
> I'm afraid this has broken dead keys (at least) on macOS: typing `´e` now inserts `éé` instead of a single `é` as on main (and similarly for other dead key combinations).
> 
> Also, typing `´´` (or other dead keys) makes dead keys "sticky", so any other dead key is inserted literally _twice_, and this continues while typing (other) dead keys.

as is mentioned before, https://github.com/neovide/neovide/pull/1083#issuecomment-991083715, this line code, i think it's not a right way, but i don't know what this code mean, so i can't fix this. But i think this is another bug hidden before, and it's  bring up after i fix the ime input.

**Comment by clason** [comment]
Just to make clear: dead keys are working just fine on the main branch; this is a definite regression with this PR.

**Comment by clason** [comment]
Protip: don't paste screenshots but links to the file through the Github API, like so: https://github.com/neovide/neovide/blob/7b6a8f31d2de1fca52a5cb039b97f5c7f92e5cbd/src/window/keyboard_manager.rs#L104-L108

This makes it easier for people to check the context (especially blame).

**Comment by vzex** [comment]
> Protip: don't paste screenshots but links to the file through the Github API, like so:
> 
> https://github.com/neovide/neovide/blob/7b6a8f31d2de1fca52a5cb039b97f5c7f92e5cbd/src/window/keyboard_manager.rs#L104-L108
> 
> This makes it easier for people to check the context (especially blame). Here's the rationale explained: #767 (@smolck)

ok, i'll reproduce this later, I never used dead key before.

**Comment by vzex** [comment]
> I'm afraid this has broken dead keys (at least) on macOS: typing `´e` now inserts `éé` instead of a single `é` as on main (and similarly for other dead key combinations).
> 
> Also, typing `´´` (or other dead keys) makes dead keys "sticky", so any other dead key is inserted literally _twice_, and this continues while typing (other) dead keys.

please check, I've tested on my mbp, it's ok now.

**Comment by clason** [comment]
Sorry, now dead keys work (partly -- typing a dead key and a non-composing key swallows the latter: `´b`  should result in a literal `´b`, but you only get `´`), but Alt-modifiers are broken on non-US keyboards: Alt-8 should (and used to) insert `{`, but now I only get `8`.

**Comment by vzex** [comment]
> 

Alright, I'll have a try, there are so many special cases.

**Comment by clason** [comment]
> Alright, I'll have a try, there are so many special cases.

Yes indeed; I'm afraid there's no alternative to a lot of special cases depending on platform and language. Any attempt to "simplify" things _will_ break one or more use case in my experience. 

(The core of a GUI is input and output -- which otherwise a terminal will handle for you. People tend to focus on the output part, since that's where "the shiny" is, but the input part is just as important and arguably harder; especially if you want to be cross-platform without pulling in a huge framework like QT...)

**Comment by vzex** [comment]
> > Alright, I'll have a try, there are so many special cases.
> 
> Yes indeed; I'm afraid there's no alternative to a lot of special cases depending on platform and language, and supporting IME necessarily means adding another. Any attempt to "simplify" things _will_ break one or more use case in my experience.
> 
> (The core of a GUI is input and output -- which otherwise a terminal will handle for you. People tend to focus on the output part, since that's where "the shiny" is, but the input part is just as important and arguably harder; especially if you want to be cross-platform without pulling in a huge framework like QT...)

I tested on windows and mbp, with chinese input, dead keys(alt+e u, alt+8(with us keyboard and austria keyboard layout)).All works well now(just for now).

**Comment by vzex** [comment]
> > Alright, I'll have a try, there are so many special cases.
> 
> Yes indeed; I'm afraid there's no alternative to a lot of special cases depending on platform and language, and supporting IME necessarily means adding another. Any attempt to "simplify" things _will_ break one or more use case in my experience.
> 
> (The core of a GUI is input and output -- which otherwise a terminal will handle for you. People tend to focus on the output part, since that's where "the shiny" is, but the input part is just as important and arguably harder; especially if you want to be cross-platform without pulling in a huge framework like QT...)

Check again please

**Comment by clason** [comment]
Yes, now modifier keys work again 🚀 

(The `´b` only printing `´` regression is still there, but that's minor if it means adding support for IME.)

**Comment by vzex** [comment]
> Yes, now modifier keys work again 🚀
> 
> (The `´b` only printing `´` regression is still there, but that's minor if it means adding support for IME.)
 
I 've solve the problem now ,but I found a new problem, that is:
if I am in normal mode, when i am typing alt+q, I bind a hot-key to close buffer, but in the insert mode, the alt+q means show a character 'œ'. So I can't directly decide to send '<A-q>' or 'œ' before I know the vim mode. I think after bring in the ime support, the keyboardmgr should know what mode the vim in or just send the keyevent to 

> Yes, now modifier keys work again 🚀
> 
> (The `´b` only printing `´` regression is still there, but that's minor if it means adding support for IME.)

I have added more support for dead keys, and list every type of cases.

**Comment by clason** [comment]
> I have added more support for dead keys, and list every type of cases.6 (Austria keyboard):[= =] =>´´q, this means [= = q] =>´´q

Yes, that works now, as well. (I don't use `alt` keybindings, since terminals don't allow them anyway, sorry...)

**Comment by vzex** [comment]
> if I am in normal mode, when i am typing alt+q, I bind a hot-key to close buffer, but in the insert mode, the alt+q means show a character 'œ'. So I can't directly decide to send '' or 'œ' before I know the vim mode. I think after bring in the ime support, the keyboardmgr should know what mode the vim in or just send the keyevent to
> 
> >

Never mind, I found that Neovide didn't support alt on mac yet, so my pr doesn't break this future

**Review by last-partizan — CHANGES_REQUESTED** [review]

File: src/window/keyboard_manager.rs [hunk 3]
```diff
@@ -8,14 +9,19 @@ use crate::channel_utils::LoggingTx;
...
+enum InputEvent {
+    KeyEvent(KeyEvent),
+    ImeInput(String),
+}
 pub struct KeyboardManager {
     command_sender: LoggingTx<UiCommand>,
     shift: bool,
     ctrl: bool,
     alt: bool,
+    pre_is_dead: Option<char>,
```
[the remark this answers is not in the document]
Comment (vzex) [reply src/window/keyboard_manager.rs:21]: agree

**Comment by vzex** [comment]
So is there anyone can approval this?

File: src/window/keyboard_manager.rs [hunk 4]
```diff
@@ -69,22 +84,42 @@ impl KeyboardManager {
...
-                                    .send(SerialCommand::Keyboard(keybinding).into())
-                                    .expect("Could not send keyboard ui command");
+                    let mut pre_is_dead = self.pre_is_dead;
+                    for input_event in self.queued_input_events.iter() {
+                        match input_event {
+                            InputEvent::KeyEvent(key_event) => {
+                                // And a key was pressed
+                                if key_event.state == ElementState::Pressed {
+                                    if let Some(keybinding) =
+                                        self.maybe_get_keybinding(key_event, &mut pre_is_dead)
```
[the remark this answers is not in the document]
Comment (last-partizan) [reply src/window/keyboard_manager.rs]: @vzex this is unresolved, prev_dead_key should not be passed as mutable here. Or there is a reson why this impossible without passing prev_dead_key?

**Comment by vzex** [comment]
@last-partizan I've uploaded the new changes, it's now  don't need to pass the prev_dead_key

**Review by last-partizan — APPROVED** [review]

**Comment by last-partizan** [comment]
@Kethku i think this code is in better shape that it was before, and it's ready for your review.

**Comment by Kethku** [comment]
Awesome. Thanks for the ping. I will take a look

File: src/window/keyboard_manager.rs [hunk 5]
```diff
@@ -97,19 +132,27 @@ impl KeyboardManager {
...
+            }
+        } else {
+            let key_text = if self.prev_dead_key.is_none() {
                 key_event.text
+            } else {
+                key_event.text_with_all_modifiers()
             };
-
-            if let Some(key_text) = key_text {
+            if let Some(ori_key_text) = key_text {
```
Comment (Kethku) [inline src/window/keyboard_manager.rs:149]: Super minor, where does the name `ori_key_text` come from?

Comment (Kethku) [reply src/window/keyboard_manager.rs:149]: is it original?

**Comment by Kethku** [comment]
This looks good to me! Thanks for the amazing persistence. This has been an open issue for a while. Once the lint build comes back I will merge

File: src/window/keyboard_manager.rs (hunk 5 above)
Comment (last-partizan) [reply src/window/keyboard_manager.rs:149]: Looks like "original" to me, i'll fix it in main branch :)

**Pull request merged**